chains/ethereum/gnosis_forwarder_client.go

```diff
@@ -0,0 +1,227 @@
+
+var ZeroAddress = common.HexToAddress("0x0000000000000000000000000000000000000000")
+
+var gnosisForwarderAbi abi.ABI
+
+func init() {
+	var err error
+	gnosisForwarderAbi, err = abi.JSON(strings.NewReader(GnosisForwarderAbi))
+	if err != nil {
+		panic(err)
+	}
+}
+
+func NewGnosisForwarderClient(
+	client *ethclient.Client,
+	forwarderAddress common.Address,
+	fromAddress common.Address,
+	chainId *big.Int,
+) *GnosisForwarderClient {
+	return &GnosisForwarderClient{
+		ForwarderClientBase{
+			client:           client,
+			forwarderAddress: forwarderAddress,
+			fromAddress:      fromAddress,
+			forwarderAbi:     gnosisForwarderAbi,
+			chainId:          chainId,
+		},
+	}
+}
+
+func (c *GnosisForwarderClient) GetOnChainNonce() (*big.Int, error) {
+	packed, err := c.forwarderAbi.Pack("nonce")
+	if err != nil {
+		return nil, err
+	}
+
+	data := eth.CallMsg{
+		To:    &c.forwarderAddress,
+		Data:  packed,
+		From:  c.fromAddress,
+		Gas:   200000,
+		Value: big.NewInt(0),
+	}
+
+	callOpts := &bind.CallOpts{From: c.fromAddress, Context: context.Background()}
+	res, err := c.client.CallContract(callOpts.Context, data, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := c.forwarderAbi.Unpack("nonce", res)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := p[0].(*big.Int)
```

Comment:
You could do the following to check that the type assertion actually succeeded, otherwise you will panic in an instance when it doesn't. Not sure if that could happen considering you check the error above, but anyway:

```suggestion
	nonce, ok := p[0].(*big.Int)
	if !ok {
	  // handle it
	}
```

Reply:
Good point, updated